llpc/util/llpcUtil.h

```diff
@@ -114,4 +115,26 @@ inline unsigned log2(T u) {
   return logValue;
 }
 
+// Returns true if shaderInfo has the information required to compile an unlinked shader of the given type.
+bool hasDataForUnlinkedShaderType(Vkgc::UnlinkedShaderStage type,
+                                  llvm::ArrayRef<const Vkgc::PipelineShaderInfo *> shaderInfo);
+
+// Returns the shader stage mask for all shader stages that can be part of the given unlinked shader type.
+unsigned int getShaderStageMaskForType(Vkgc::UnlinkedShaderStage type);
+
+// Returns the name of the given unlinked shader type.
+const char *getTypeName(Vkgc::UnlinkedShaderStage type);
+
+// Pre-increment operator on the unlinkeed shader type.
+inline Vkgc::UnlinkedShaderStage operator++(Vkgc::UnlinkedShaderStage &type) {
```

Comment:
I guess the alternative would be to have something like the enum iterator in LLVM: https://github.com/llvm/llvm-project/blob/1465e7770bcacfb67ca4edc3714cf3a7e8a2a110/llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp#L56. This could be added to some utility header in LLPC. This seems like a cleaner abstraction to me compared to loops with casts and these increments, but I don't have a strong preference.

Reply:
This will be up to AMD, but we should probably get a standard solution in place.  I do not like the casts in the loops either.  To remove the cast for the initialization, the best way is to introduce a `_FIRST` element, but I was asked to remove it (https://github.com/GPUOpen-Drivers/llpc/pull/1264#discussion_r646620996).  I don't  like the template because not all enumerated types can be iterator over that way, and it could be easy to think you are doing something correct when you are not.  I like having those at can be explicitly marked as such.  We could also declare an array containing all of the enums and iterator over the array.

Reply:
We could mark iteratable enums with a tag and only accept that in the utility implementing enum ranges. Something like:
```
// In a utility header:
template <typename E>
struct iteratable_enum;

template <typename E, typename = std::void_t<iteratable_enum<E>>
struct enum_range {

}; ...

// At the enum definition place.
enum MyEnum : unsinged {
X = 0,
...,
Z,
Last = Z
};

template <>
struct iterable_enum<MyEnum> {};
```

Reply:
But anything more complicated than these increment operators should probably go into a separate PR anyway

Reply:
Agreed, and I see you open something up for discussion.  I'll mark this conversation as resolved.